CORDA-1787 Bring threat model into the docsite

jamesbr3 opened the pull request:
The threat-model has been updated, refreshed and formatted to look more like a threat model. 

The content has been copied across (auto-formatted) from:
https://r3-cev.atlassian.net/wiki/spaces/SEC/pages/561316232/Corda+Threat+model+interim

and will appear under:

docs/source/threat-model/

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+Corda employs a Public Key Infrastructure (PKI) to validate the identity of
+nodes, both at the point of registration with the network map service and
+subsequently through the cryptographic signing of transactions. An imposter
+would need to acquire an organisation's private keys in order to meaningfully
```
They could also trick the doorman into issuing an incorrect certificate, right? Which leads to a discussion of a plausible end-state vision where, to mitigate the risk of incorrectly issued certificates, we highlight that it is the responsibility of a business network operator only to whitelist (for whatever definition is meaningful in the context of their app(s)) certificates that they have independently confirmed have been issued correctly.

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+would need to acquire an organisation's private keys in order to meaningfully
+impersonate that organisation. R3 provides guidance to all ledger network
+participants to ensure adequate security is maintained around cryptographic
+keys.  
```
... and we are working on HSM support for nodes. Plus the longer-term direction for client-side signing devices in the whitepaper.

jamesbr3 replied:
good point, i've updated to mention hsm here

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+attempting to spoof either service would need access to the corresponding
+private keys.
+
+The Doorman and NetworkMap signing keys are stored inside a (Hardware Security
```
clarify that the key used to sign the netmap is different to the key used to secure the TLS session?

jamesbr3 replied:
done

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+The node's vault is assumed to be within the same trust boundary of the node.
+Access to the vault must be restricted such that only the node can access it.
+Both network-level controls (fire-walling) and database permissions must be
+employed.
```
elaborate on the point that the only party this harms is the party whose vault it is, right?  ie nobody else can be fooled.

jamesbr3 replied:
very good point; updated

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+limited in this regard.
+
+The only impact the attacker could have would be to remove individual entries
+in the map, or insert their own NodeInfo.
```
perhaps elaborate further on the role of the additional-node-infos directory?  ie even in the event of a total network map server outage (perhaps this should be under denial of service below), it  _would_ be possible, if annoying to distribute NodeInfos out of band?

jamesbr3 replied:
this is covered in the denial-of-service section so i'll leave this as-is

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+Preventing repudiation does not prevent other forms of attack. Rather, the
+goal is to ensure that the attacker is identifiable, their actions can be
+traced, and there is no way for the attacker to deny having committed those
+actions.  
```
these two paragraphs are a little confusing.  Perhaps it's my lack of understanding but my mental model (until now!) of repudiation was that it 1) ENABLES me to prove/deny something that didn't (correctly) happen actually didn't happen and 2) PREVENTS me from denying something that DID happen actually happened.

jamesbr3 replied:
i'll try and clarify this, obviously this part needs work. 

Repudation is an attacker's ability to deny their actions. Preventing repudation is therefore providing some kind of a defence. i'll fix

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+</td>  
+<td>
+
+Attacker attempts to initiate a transcation they are not entitled to perform
```
typo

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+</td>  
+<td>
+
+All P2P transactions are logged by the node.
```
this section seems a bit light.     And isn't the thing that prevents me from spending something that doesn't belong to me the verification logic of the associated transaction?

jamesbr3 replied:
noted. i'll try and include some better descriptions here. Repudiation is a difficult one to describe :D

mikehearn replied:
Yeah, isn't the thing that prevents repudiation here the signatures?

We could also mention that non-validating notaries require a signature over the request so nobody can deny doing a denial-of-state attack.

jamesbr3 replied:
i've included TLS authentication & transaction signing as mitigations

gendal commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+</td>  
+<td>
+
+A malicious CordApp attempts to spend a transaction that does not belong to
```
CorDapp (capital D).  Also - what does it mean to spend a transaction. Do you mean a state?

jamesbr3 replied:
yes, i meant state. fixed

gendal commented on docs/source/threat-model/corda-threat-model.md:
Is that correct?  I thought the current implementation (design?) is such that a node will only divulge a tx history to a peer if the tx being asked for is a descendent of a tx that the node itself had previously shared as part of the current session - or something like that.

jamesbr3 replied:
i need to think about this!

mikehearn replied:
yes richard is right - you can request any tx in the transitive dependency graph of a transaction. It must be so otherwise you couldn't check tx validity.

jamesbr3 replied:
updated & clarified

gendal commented on docs/source/threat-model/corda-threat-model.md:
I think the risk of guessing has been mitigated through the "you can only request txs or their histories if they were sent to you by the node you're asking" feature (I think!).  Andrius implemented it IIRC

jamesbr3 replied:
yes ok. i copy-pasted as much as I could from the previous threat-model document - this bit was clearly out-of-date. fixed

gendal commented on docs/source/threat-model/corda-threat-model.md:
its

gendal commented on docs/source/threat-model/corda-threat-model.md:
do we need to talk about how the credentials used by the node to connect to the database or HSM are protected? ie by default in a plain text file but we have mechanisms to inject credentials via env vars, etc?

also - how is the float/bridge protected? especially the bridge key material.

gendal commented on docs/source/threat-model/corda-threat-model.md:
and notaries?  or do we regard them as special cases of nodes?

jamesbr3 replied:
I haven't drawn any distinction actually. Arguably I should but I was thinking to leave that to a separate higher-level 'ledger' threat model rather than give too much detail here. But I think it is worth mentioning, I will make this change, but not go into too many  specifics

gendal commented on docs/source/threat-model/corda-threat-model.md:
their?

jamesbr3 replied:
fixed

gendal commented on docs/source/threat-model/corda-threat-model.md:
rather than specify the time maybe just refer to the event horizon concept?

gendal commented on docs/source/threat-model/corda-threat-model.md:
CorDapps

gendal commented on docs/source/threat-model/corda-threat-model.md:
this is a critical point... perhaps important enough to elevate to the introduction?  ie it's a departure from the threat model of bitcoin/ethereum (whether they achieve resilience to the threat is another question) but it also directly leads to business/commercial decisions about how widely available to make the testnet, etc...     should we want to be more resilient against nOn attacks, what would we do?

gendal replied:
ah - I see... you answer that last question on the next line :)

jamesbr3 commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,883 @@
+Paper](https://docs.corda.net/_static/corda-technical-whitepaper.pdf) for a
+detailed description of Corda's design and functionality.
+
+R3 will provide and maintain a number of essential services underpinning the
```
ok updated

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+
+Built on the [Corda](http://www.corda.net/) distributed ledger platform
+designed by R3, the ledger network enables the origination and management of
+(primarily financial) agreements between business partners. Participants to
```
Can we take out the (primarily financial) part? We're getting quite a lot of non financial firms these days.

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+separate Corda Foundation. The network services currently include:
+
+  * Network map service: Service for distributing and providing updates for a network map document enabling nodes to identify other nodes on the network, their network address and advertised services. 
+  * Network permissioning service ('Doorman'): Issues signed digital certificates credentialising the identity of parties on the network to conduct peer-to-peer communication.
```
credentialising? :)

jamesbr3 replied:
i copy-pasted from the old document :D

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+# The Threat Model
+
+Threat Modelling is an iterative process that works to identity, describe and
+mitigate threats to a system. One of the most common models for identify
```
typo: identify -> identifying

jamesbr3 replied:
fixed

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+identity.
+
+An attacker attempts to impersonate another node on the network by submitting
+NodeInfo updates with - falsified address and/or identity information.
```
do we need the hyphen here?

jamesbr3 replied:
fixed

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+order to meaningfully impersonate that organisation. Corda will soon support a
+range of HSMs (Hardware Security Modules) for storing a node's private keys,
+which mitigates this risk.
+
```
Do we need to discuss the case where the zone operator is the attacker here? The operator can issue an identity cert for anyone, publish a valid NodeInfo and redirect all traffic to themselves in the clear.

jamesbr3 replied:
good point. i've clarified this in the 'doorman' element further down

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+
+Mutually authenticated TLS connections between nodes ensures that Man-In-The-
+Middle (MITM) attacks cannot take place. Corda Nodes restrict their
+connections TLS v1.2 and also restrict which cipher suites are accepted.
```
connections _to_ TLS 1.2

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -0,0 +1,889 @@
+limited in this regard.
+
+The only impact the attacker could have would be to remove individual entries
+in the map, or insert their own NodeInfo.
```
The operator could change a NodeInfo by also issuing a new identity cert, right? There's nothing in the node today that flags unexpected key changes.

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
This sounds a bit vague. Perhaps we can say, network map is secured by CDNs and large-scale anti-DDoS networks. Notaries and oracles will only communicate with peers in the network, so are protected from non-member-on-member app level attacks.

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
should flow be all lower case here?

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
typo: Doorman

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
that *installed* CorDapps are trusted. We must be careful to distinguish here between contracts/states and flows.

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
We *do* assume peers may be sending malformed data. We *don't* trust other network participants.

jamesbr3 replied:
Hmm OK I was under the impression that the current design was never intended to mitigate this. But I agree, it should be part of the threat model - i'll remvoe this part

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
SGX provides privacy but not security - it stops code getting in not out.

jamesbr3 replied:
good point, i'll remove this from the EoP section - not relevant

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
This is a good document. I'd like to see some aspects of the original document in perhaps a separate section, in particular enumerating what attacks each class of participant are assumed to be capable of. For instance what can and cannot the zone operator do? What about peers? What about notaries? etc. People need to be able to do reasoning of the form "if company X goes bad what's the damage"? The STRIDE framework seems like a good way to force thinking about different classes of attack, but I suspect from a business level security reasoning will be people oriented rather than attack class oriented.

mikehearn commented:
Also can you edit the toc/index files so it will be rendered as part of the docsite? Not sure which section it should go under ....

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -228,14 +228,15 @@ Doorman
 
-An attacker attempts to join the Corda Network by impersonating an existing
-organisation and issues a fraudulent registration request
+An malicious zone operator attempts to join the Corda Network by impersonating
```
I think "attacker" was right before. Zone operators can't join zones, they run them.

jamesbr3 replied:
fixed

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -379,15 +380,16 @@ or removed from the map.
-modifications to a NodeInfo. It is not possible for the NetworkMap service (or
-R3) to modify the NodeInfo entries, therefore an attacker would also be
-limited in this regard.
+Node's private key. The signatures are validated by the NetworkMap service,
```
node to lowercase

jamesbr3 replied:
fixed

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -632,11 +695,12 @@ availability of targeted ledger network node(s)/service(s), both during the
-enterprise-grade DDoS detection and mitigation services, which will conduct
-intelligent traffic profiling in order to detect and block malicious traffic.  
+network as a whole is relatively resilient to DoS attacks. Notaries and
+Oracles will only communicate with peers in the network, so are protected from
```
oracles to lower case

jamesbr3 replied:
fixed

mikehearn commented on docs/source/threat-model/corda-threat-model.md:
```diff
@@ -763,9 +826,11 @@ Updates to the network map must be signed by participant nodes and are
 
 The network map is designed to be distributed by a CDN (Content Delivery
-Network).
+Network). It is secured large-scale anti-DDoS networks.
```
This sentence isn't grammatical

jamesbr3 replied:
fixed

mikehearn commented:
Also please remember to link it into the docsite toctree, and do a local re-render to ensure it looks good.

jamesbr3 commented:
I've updated typos and have moved the threat model under the Design section of the docsite (and updated the toc). Just need to get it rendered locally
